Avoid caching compilation data and use value equality for SyntaxNodes

jtschuster opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/78242.

Creates separate types for InteropAttributeData: one that holds compilation data (InteropAttributeCompilationData), and one that holds only the data necessary for the model to create the generated code (InteropAttributeModelData).

This uncovered some issues with record equality in records that use SyntaxNode. For those, we need to override Equals or wrap the SyntaxNode in a type that overrides Equals to use IsEquivalentTo on the SyntaxNode.

There are probably more places where we use SyntaxNode that aren't caught in the current tests.

To make sure every record has the right equality, I wasn't sure if it would be better to override Equals for each of the records, or create a wrapper record struct for each SyntaxNode that implements the equality we want (and implicit casts to and from the SyntaxNode). Then we wouldn't have to explicitly override the equality in each record that has a SyntaxNode. I also overrode both Equals and GetHashCode, but I'm not confident in my GetHashCode implementation. It could also be done with IEquatable.Equals without needing GetHashCode, but that would require implementing the TypeSyntax equality for every type that inherits from ManagedTypeInfo.

jtschuster committed:
Avoid caching compilation data and use value equality for SyntaxNodes

teo-tsirpanis commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs:
```diff
@@ -24,27 +23,30 @@ public enum InteropAttributeMember
     /// Common data for all source-generated-interop trigger attributes
     /// </summary>
-    public record InteropAttributeData
+    public record InteropAttributeCompilationData : InteropAttributeModelData
```
You should also change the name of the file.

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs:
```diff
@@ -1,17 +1,27 @@
 namespace Microsoft.Interop
 {
     /// <summary>
     /// LibraryImportAttribute data
```
Doc is out of date.

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/LibraryImportData.cs:
```diff
@@ -1,17 +1,27 @@
+        public string EntryPoint { get; init; }
+    }
+
+    internal sealed record LibraryImportData(string ModuleName) : InteropAttributeModelData
```
Type doc?

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/InteropAttributeData.cs:
```diff
@@ -24,27 +23,30 @@ public enum InteropAttributeMember
+    {
+        public INamedTypeSymbol? StringMarshallingCustomType { get; init; }
+    }
+    public record InteropAttributeModelData
```
Doc for type.

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManualTypeMarshallingHelper.cs:
```diff
@@ -24,6 +24,22 @@ namespace Microsoft.Interop
+        public bool Equals(CustomTypeMarshallers other)
+        {
+            return Modes.Count == other.Modes.Count
+                && !Modes.Except(other.Modes).Any();
```
Took me a second to understand this check. Please drop a comment on this logic.

jkoritzinsky commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs:
```diff
@@ -18,6 +15,19 @@ public abstract record ManagedTypeInfo(string FullTypeName, string DiagnosticFor
+
+        public override int GetHashCode()
+        {
+            return Syntax.GetHashCode() ^ FullTypeName.GetHashCode() ^ DiagnosticFormattedName.GetHashCode();
```
I think that the `Syntax.GetHashCode()` call here should be dropped. It doesn't look like `SyntaxNode` overrides `GetHashCode()`, so we'll end up with mismatching hash codes for equal values, which breaks a lot of things.

```suggestion
            return FullTypeName.GetHashCode() ^ DiagnosticFormattedName.GetHashCode();
```

jtschuster committed:
PR Feedback:

- Add type doc comments for new types
- Don't use SyntaxNode in GetHashCode implementation
- Add comment on Dictionary equality check

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into NoRefToComp

elinor-fung reviewed (approved):
> This uncovered some issues with record equality in records that use SyntaxNode. For those, we need to override Equals or wrap the SyntaxNode in a type that overrides Equals to use IsEquivalentTo on the SyntaxNode.
>
>There are probably more places where we use SyntaxNode that aren't caught in the current tests.

Do you think it is worth filing an issue to do a pass over this?

jtschuster commented on src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/ManagedTypeInfo.cs:
```diff
@@ -18,6 +15,19 @@ public abstract record ManagedTypeInfo(string FullTypeName, string DiagnosticFor
         private TypeSyntax? _syntax;
         public TypeSyntax Syntax => _syntax ??= SyntaxFactory.ParseTypeName(FullTypeName);
 
+        public virtual bool Equals(ManagedTypeInfo other)
```
There shouldn't be a null check, thanks for pointing that out!

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into NoRefToComp

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into NoRefToComp

jtschuster committed:
Add null check

jtschuster committed:
Add comment to not keep compilation alive

jkoritzinsky commented:
Only failure is #79874
